src/compiler/checker.ts

```diff
@@ -7629,8 +7629,12 @@ namespace ts {
             // is no 'React' symbol in scope, we should issue an error.
             if (compilerOptions.jsx === JsxEmit.React) {
                 let reactSym = resolveName(node.tagName, "React", SymbolFlags.Value, Diagnostics.Cannot_find_name_0, "React");
-                if (reactSym) {
+                if (reactSym && reactSym !== unknownSymbol) {
                     getSymbolLinks(reactSym).referenced = true;
+                    let reactNode = <Identifier>createSynthesizedNode(SyntaxKind.Identifier, false);
+                    reactNode.text = 'React';
+                    getNodeLinks(reactNode).resolvedSymbol = reactSym;
+                    node.reactNode = reactNode;

```

Comment:
i do not think this is going to work with incremental parsing. now we are adding artifacts on the node that will not be removed if the user removed react.

I think the right way to handle this is for the emitter to ask the checker or the emit resolver to be more accurate about it, and the checker could possibly cache it if needed in the nodeLinks, though might not be necessary. 

check with @ahejlsberg as well.